pkg/devicetemplateconverter/profile_fetcher.go

```diff
@@ -0,0 +1,57 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicetemplateconverter
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/cluster"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"google.golang.org/grpc"
+)
+
+// ProfileFetcher handles the validation and fetching of end device profile in the Device Repository.
+type ProfileFetcher struct {
+	GetPeerConn func(ctx context.Context, role ttnpb.ClusterRole, ids cluster.EntityIdentifiers) (*grpc.ClientConn, error)
+	CallOpts    func(context.Context) ([]grpc.CallOption, error)
+}
+
+// ValidateFields dictactes if the end device has the necessary fields to fetch its profile.
+func (pf *ProfileFetcher) ValidateFields(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetVendorId() != 0 &&
+		device.GetVersionIds().GetVendorProfileId() != 0
+}
+
+// FetchProfile provides the end device profile.
+func (pf *ProfileFetcher) FetchProfile(ctx context.Context, device *ttnpb.EndDevice) (*ttnpb.EndDeviceTemplate, error) {
+	conn, err := pf.GetPeerConn(ctx, ttnpb.ClusterRole_DEVICE_REPOSITORY, nil)
+	if err != nil {
+		log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
+		return nil, err
+	}
+	opts, err := pf.CallOpts(ctx)
```

Comment:
Wouldn't we naturally forward the auth here anyway?

Reply:
It is called but not so by conventional means, [here](https://github.com/TheThingsNetwork/lorawan-stack/blob/cc4e6e5928cb0adbb4c6a8959c915ea88265dd6b/pkg/devicetemplateconverter/devicetemplateconverter.go#L55-L61) its where the callOpts for the profile fetcher are specified. I tried to make something a bit more customization in order to facilitate testing of the fetcher. This was something that I should've mentioned in the notes for reviewers, since it is not the standard way of doing so. If you feel it is inappropriate I'll revert it.

Reply:
The way we do this is by defining the target client. See https://github.com/TheThingsNetwork/lorawan-stack/blob/cb215d00b2f1563f9620c653a0ed223025de71ff/pkg/deviceclaimingserver/enddevices/enddevices.go#L254-L260

I don't think adding the `callopts` to the constructor is the correct way to go. With the suggested method, the mock just needs to implement the client interface functions instead of needing to be an RPC server.